### DISCUSSION TITLE
Allow editing of Radio station URLs and images

### DISCUSSION THREAD

**Discussion by dunxd** [music-assistant/support#3279] [Feature requests and ideas]
At the moment if an error is made in entering details for for a radio station the only way to fix this is to delete the station and add it again.  It would be nice to be able to edit all the properties of radio stations - stream URL, custom name and image URL - in case of errors or the URLs changing.  

Since the station name might be linked to automations etc. to make changes properly means adding a new station with the correct details but a different name, confirming it works, then deleting both the new and original stations and adding it back again with the name used in the automations.

**Comment by OzGav** [comment]
On your second point you should be using the station name in your automations so this is moot.

**Reply by dunxd** [reply]
I don't understand what you mean.

I _am_ using the station name in my automations.  But you can't have two stations with the same name right?  So if you have to correct any details you have to create a new station with a different name, and then either update all automations to match that or follow the procedure I described.

When I say "station name" I mean the box labelled "Enter a new (custom) name".  Which I normally change, because the names provided in the stream metadata are usually ugly or confusing.  Automations work with this so I am assuming that is what you mean.

**Reply by OzGav** [reply]
No you would delete the station that needs updating then add it back with the same name.

**Reply by dunxd** [reply]
That is one option for doing things at the moment.  If updating was possible it would be easier and quicker to make minor changes to URLs etc.

At the moment it isn't possible to retrieve all the existing values - the station image for example, so making minor adjustments is very fiddly.

Some URLs encode the quality and you might want to tweak that, and the BBC regularly changes the official URLs for their streams, so making changes isn't that rare of an occurrence for me.

**Reply by cpwom** [reply]
This program is sooo cool it's true that editing a station should be a trivial must have. Same problem here ... some station have (or even don't have) poor images .. and I create some for a radio wall I have. So sometimes I need editing to change the image.

**Reply by dunxd** [reply]
BBC keep changing their radio feed URLs, meaning this is a fairly regular task for me. 

The built in radio station database has hundreds of outdated and broken versions of BBC radio feeds that look identical in the interface, so manually adding them is the best solution. 

The ability to edit would be a regular benefit for me.

**Reply by vegancodecruncher** [reply]
> BBC keep changing their radio feed URLs, meaning this is a fairly regular task for me.
> 
> The built in radio station database has hundreds of outdated and broken versions of BBC radio feeds that look identical in the interface, so manually adding them is the best solution.
> 
> The ability to edit would be a regular benefit for me.

One approach would be to run a local DNS server and create a CNAME for each station. You would then reference that in MA and could programmatically update the DNS record :)

**Comment by dunxd** [comment]
A CNAME maps a host name to another host name. BBC have a lot of stations that are accessed through the same hostname with different URLs. CNAME is not a viable solution. Consider this:

`http://hostname.com/resource.mp3` changes its URL to `http://newhostname.con/resources.mp3`
CNAME can fix this as it can remap `hostname.com`.

But if `resources.mp3`  changes at all, a CNAME DNS record can't fix that.

Even if it did work, how many MA users also have a DNS server they can manage. I would assume even for the people that want to edit URLs the Venn overlapp is tiny.

~~Coincidentally, i think BBC broke all the links again in the last week.~~ Ignore that. It may have been a glitch this morning.

**Reply by cpwom** [reply]
And just to change the image of the station ... let's do it with a CNAME ...  I'm really surprised that letting modifying the name, url or image seems so difficult to do.

**Reply by OzGav** [reply]
Why do you think it is difficult to do?

**Reply by dunxd** [reply]
Are you asking if it is currently difficult to change image, name,.URL for radio stations in Music Assitant - see my original that describes the process which I understand is the only option currently and I consider difficult. Not the highest priority by any means, but I for one would find it nice.

Or are you asking if implementing the feature to allow editing of existing radio station details in Music Assistant is considered difficult?

**Reply by OzGav** [reply]
I am asking why it was said “ I'm really surprised that letting modifying the name, url or image seems so difficult to do”.

**Reply by cpwom** [reply]
well I think being able to modify the 3 properties (name, url and image) of a created radio station should be something implemented natively,  as UX is really bad without this option. It's clear that many people will need to do that often. So I can't understand it's not done for long, so maybe it's difficult, even if I can't see why.

**Reply by OzGav** [reply]
If you think hard I’m sure you can work it out. Feel free to put in a PR since you feel it is easy to do.

**Reply by cpwom** [reply]
I'm not a dev ... and be sure I'm more than thankful for this great piece of software. Didn't want to sound mean or anything negative here.

**Comment by thetechjester** [comment]
I vote a resounding yes please for this feature. MusicAssistant is awesome. Adding this feature would make the radio stations much more nice to use.... especially from a typical end user point of view. I'm no coder, so I have no idea how hard it would be to implement. Would definitely vote a YES PLEASE (5 thumbs instead of stars) :thumbsup::thumbsup::thumbsup::thumbsup::thumbsup: for this feature request.

**Reply by cpwom** [reply]
I've recently found a workaround ... If you simply add a radio station with the exact same name than the one you wanna modify, its actually replace it with the new datas (tested it with the image url in my case). But the station ID is the same and all is fine.

**Comment by Borrower543** [comment]
+1 for this feature. Some radio stations import with generic names like "index" or a string of numbers. It would also be nice to add images.

**Reply by OzGav** [reply]
Import from where? If you add a station manually you can configure the name and  add an image.

**Comment by Borrower543** [comment]
You can't actually, the name is imported from the URL and can't be changed. If you put the URL of an image it doesn't save. Might be a bug.

-------- Original Message --------

**Reply by OzGav** [reply]
Yes it was. This will be fixed in the next nightly / beta.

**Comment by UKenGB** [comment]
Doesn't seem to be an official voting mechanism, but +100 from me. I think allowing the user to enter data but then not allowing them to edit it is terrible design.

Like others have said, Music Assistant is terrific and I am likewise very grateful to those who created it. However, it is not perfect and some things, like no editing radio station data is one of them. I have written software myself in the past and wouldn't dream of producing something with such a glaring omission. Was this by design, or just a mistake?

As I said, not trying to defame MA in general, but this is an appalling lack of functionality.

Trying to set up radio station entries, that work and keep working is proving a nightmare and although the problems are not of MA's making, MA is making the whole problem worse by not providing an editing interface for the radio station data. I think a single entry is split across 2 tables, but this is trivial SQL to edit the data. I'd do it myself, but MA locks the entire SQLite database so could only be done by shutting down MA which is not practical every time a minor edit is required.

MA is great, but sorry, this particular aspect is dreadful. Please, I implore you to add the facility to edit radio station date.

**Reply by OzGav** [reply]
There is an offical voting mechanism. People upvote the first post of the feature request. This feature request is way down the list and that is from a request made in Dec 2024 so your "glaring omission" doesn't seem so critical to the rest of the community. I have mentioned this to you before but your tone is not encouraging us to dive in and help. We are a small team so this is about workload. We work on things that the community wants as voted by them.

**Reply by Borrower543** [reply]
+1 from me, surprised it's not a feature.

**Reply by UKenGB** [reply]
In reply to @OzGav 

Hence why I did vote on the first post.

I'm sorry you find my comments discourage you from helping, but I have tried to make it clear that I believe MA is a terrific product. However, that doesn't mean it's perfect and I do have some experience in the matter of UI development and quite frankly it is inexplicable to me that you seek to justify such a terrible piece of UI design. The fact that few others are prepared to try and make this clear to you doesn't make it 'ok' and therefore not important. Most users will never see this 'forum' so you'll never hear of their MA grievances.

I am fully aware you are a small team and if your response was, "yes we know but we're snowed under and will try to sort this as soon as we can make the time" I would be perfectly satisfied. Instead, you try and justify such poor design and that is, let's just say, less satisfactory.

I say again, MA is terrific and I applaud all your efforts (including you for the time and effort you put in to its support), but you need to recognise where there is need for improvement - like when those users prepared to  make the effort are trying to tell you. 😀

**Reply by OzGav** [reply]
You've described our work as 'terrible design,' a 'glaring omission,' an 'appalling lack of functionality,' and 'dreadful.'

While we appreciate your interest in Music Assistant, using this kind of language—even when wrapped in praise—is not constructive. As a small team, we prioritise work based on community voting. This specific request is currently low on that list.

We have noted your feedback. This is now off-topic and is not to be continued further.

**Comment by OzGav** [comment]
This will be available in 2.9.0 beta 6 except if the URL needs to be changed you must remove and re-add the station. For station name and image you will now be able to edit it. (Same goes for manually added tracks and playlists) For clarity, only manually added stations can be edited.

**Reply by UKenGB** [reply]
The ability to edit the name and logo is certainly good to have, but I would still like to be able to edit the station URL.

As it stands we are shown (read only) the station URI which is the internal MA URI. I ask that we are also shown the Station's actual URL that was entered when the station was created in MA. Editability would be fantastic, but if not, then at least display that external URL. 

The reason for this is if there are many possible source URLs for a station (there usually are) it can sometimes be important to know which URL is being used as the source. The internal URI as currently shown provides no information about the source used for the station and it is now many months since I set them all up, with no way to know the source being used. In some cases I think I want to change it (and have to create a new Station to do that), but I have no way of knowing whether that is required as I'm not shown that, IMO, crucial information.

**Reply by OzGav** [reply]
The database keys off the URL so you can't just edit it. If you have to change the URL then remove the station and add it again. It is just maybe four more clicks,

If you want to look at the original URL then look in the Provider Details section,

**Reply by disconn3ct** [reply]
4 clicks to remove and add it. How many clicks to track down every dashboard or automation that has the old MA url?

**Reply by UKenGB** [reply]
> If you want to look at the original URL then look in the Provider Details section,

Ah, got it thanks. I thought that was empty and did not notice it could be expanded and there was that exact information available there.

I had figured that you use the URL as the key, hence why it cannot be edited, but you know that's not good database design, right? You should use keys that are not part of the data/content so they never need to be edited, but all other data can be. Using the record's ID being the common practice.

As has been mentioned before, the big problem is the radio conglomerates who deliberately obfuscate the URL to their stations with changing IDs and that means constantly editing the URL to remain playable and having to create a new record every time and delete the old one is certainly more hassle than it should be for such a simple task.

Mostly I just refuse to listen to any station whose owners attempt to prevent me from listening, but it's not always so cut and dried. I'd like to think that in the future an effort could be made to restructure this aspect of the MA database to eliminate this problem.

**Reply by OzGav** [reply]
@disconn3ct zero because you shouldn’t have used the library URI. And drop the passive aggressive.

**Reply by disconn3ct** [reply]
I merely responded in kind to your clicks comment. 
If the station id isn't correct, what should be used to map a station? The action takes the id either as a URL or as a decomposed URL..

**Reply by OzGav** [reply]
https://www.music-assistant.io/faq/how-to/

**Comment by OzGav** [comment]
The feature request has been implemented successfully. As such this feature request will be closed and locked. If there are enhancements desired then open a new [feature request](https://github.com/orgs/music-assistant/discussions/categories/feature-requests-and-ideas). If there is a question about how this works then check the [documentation](https://beta.music-assistant.io/), or ask a question in the [Q&A](https://github.com/orgs/music-assistant/discussions/categories/q-a) area or on Discord. If there are problems then ask a question in the [Q&A](https://github.com/orgs/music-assistant/discussions/categories/q-a) area or on Discord or open an [issue](https://github.com/music-assistant/support/issues). Thanks for your interest in seeing this idea being brought to life.
